# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1693867993 -7200
# Parent  2cfa5ca888111ebe141baf30a15eac736b64e096
Feature: Linux systray: Avoid calling gtk_status_icon_set_tooltip_markup if app_indicator_set_tooltip_full was never called.

There are still crashes on Gnome which may be related to calling gtk_status_icon_set_tooltip_markup.

diff --git a/third_party/appindicator/app-indicator.c b/third_party/appindicator/app-indicator.c
--- a/third_party/appindicator/app-indicator.c
+++ b/third_party/appindicator/app-indicator.c
@@ -82,16 +82,17 @@ typedef struct {
     gchar                *attention_icon_name;
     gchar                *absolute_attention_icon_name;
     gchar                *icon_theme_path;
     gchar                *absolute_icon_theme_path;
     gchar                *tooltip_icon_name;
     gchar                *absolute_tooltip_icon_name;
     gchar                *tooltip_title;
     gchar                *tooltip_body;
+    gboolean              tooltip_active;
     DbusmenuServer       *menuservice;
     GtkWidget            *menu;
     gboolean              item_is_menu;
     GtkWidget            *sec_activate_target;
     gboolean              sec_activate_enabled;
     guint32               ordering_index;
     gchar *               title;
     gchar *               label;
@@ -800,16 +801,17 @@ app_indicator_init (AppIndicator *self)
     priv->title = NULL;
     priv->label = NULL;
     priv->label_guide = NULL;
     priv->label_change_idle = 0;
     priv->tooltip_icon_name = NULL;
     priv->absolute_tooltip_icon_name = NULL;
     priv->tooltip_title = NULL;
     priv->tooltip_body = NULL;
+    priv->tooltip_active = FALSE;
 
     priv->connection = NULL;
     priv->dbus_registration = 0;
     priv->path = NULL;
 
     priv->status_icon = NULL;
     priv->fallback_timer = 0;
 
@@ -1954,16 +1956,17 @@ strip_markup (const gchar *markup)
 }
 
 /* This tracks changes to the tooltip associated with the app indicator */
 static void
 tooltip_changes (AppIndicator * self, gpointer data)
 {
     GtkStatusIcon * icon = GTK_STATUS_ICON (data);
     AppIndicatorPrivate * priv = app_indicator_get_instance_private (self);
+    if (!priv->tooltip_active) return;
 
     const gchar * title = (priv->tooltip_title != NULL) ? priv->tooltip_title : priv->title;
     if (title == NULL) {
         gtk_status_icon_set_tooltip_markup (icon, NULL);
         return;
     }
     gchar * title_escaped = g_markup_escape_text (title, -1);
 
@@ -2368,16 +2371,17 @@ app_indicator_set_icon_full (AppIndicato
  */
 void
 app_indicator_set_tooltip_full (AppIndicator *self, const gchar *icon_name, const gchar *title, const gchar *body)
 {
     g_return_if_fail (APP_IS_INDICATOR (self));
     gboolean changed = FALSE;
 
     AppIndicatorPrivate * priv = app_indicator_get_instance_private (self);
+    priv->tooltip_active = TRUE;
 
     if (g_strcmp0 (priv->icon_name, icon_name) != 0) {
         g_free (priv->tooltip_icon_name);
         priv->tooltip_icon_name = icon_name ? g_strdup (icon_name) : NULL;
 
         g_free (priv->absolute_tooltip_icon_name);
         priv->absolute_tooltip_icon_name = NULL;
 
